pkg/pool-stable-phantom/test/StablePhantomPool.test.ts

```diff
@@ -1322,6 +1326,166 @@ describe('StablePhantomPool', () => {
+            // This must be a real swap - otherwise the rate changes would revert along with the query
+            const { amountOut } = await pool.swapGivenIn({
+              in: tokens.first,
+              out: tokens.second,
+              amount: amountIn,
+              from: lp,
+              recipient: lp,
+            });
+
+            // Verify the new rates are reflected in the scaling factors
+            await verifyScalingFactors(await pool.getScalingFactors());
+
+            expect(amountOut).to.not.equal(previousAmountOut);
+          });
+
+          it('joins use the new rates', async () => {
+            const previousBptBalance = await pool.balanceOf(lp);
+            const bptOut = pct(previousBptBalance, 0.18);
+            const tokenIndex = pool.bptIndex == 0 ? 1 : 0;
+            const token = tokens.get(tokenIndex);
+
+            const previousJoinResult = await pool.queryJoinGivenOut({ recipient: lp, bptOut, token });
+
+            // Now update the rates
+            await updateExternalRates();
+
+            // Verify the new rates are not yet loaded
+            expect(await pool.getScalingFactors()).to.deep.equal(previousScalingFactors);
+
+            // And join again - amountIn should be different
+            // This must be a real join - otherwise the rate changes would revert along with the query
+            const newJoinResult = await pool.joinGivenOut({ from: lp, recipient: lp, bptOut, token });
+
+            // Verify the new rates are reflected in the scaling factors
+            await verifyScalingFactors(await pool.getScalingFactors());
+
+            expect(newJoinResult.amountsIn[tokenIndex]).to.not.equal(previousJoinResult.amountsIn[tokenIndex]);
+          });
+
+          it('exits use the new rates', async () => {
+            const previousBptBalance = await pool.balanceOf(lp);
+            const bptIn = pct(previousBptBalance, 0.082);
+            const tokenIndex = pool.bptIndex == 0 ? 1 : 0;
+            const token = tokens.get(tokenIndex);
+
+            const previousExitResult = await pool.querySingleExitGivenIn({ from: lp, bptIn, token });
+
+            // Now update the rates
+            await updateExternalRates();
+
+            // Verify the new rates are not yet loaded
+            expect(await pool.getScalingFactors()).to.deep.equal(previousScalingFactors);
+
+            // And exit again - amountOut should be different
+            // This must be a real exit - otherwise the rate changes would revert along with the query
+            const newExitResult = await pool.singleExitGivenIn({ from: lp, bptIn, token });
+
+            // Verify the new rates are reflected in the scaling factors
+            await verifyScalingFactors(await pool.getScalingFactors());
+
+            expect(newExitResult.amountsOut[tokenIndex]).to.not.equal(previousExitResult.amountsOut[tokenIndex]);
+          });
+
+          it('recovery mode exits do not update the cache', async () => {
```

Comment:
We should also have a test in which one of the rate providers reverts, but recovery mode exits still work.

Reply:
We do the "with broken pool" test. It doesn't actually call the rate update and check that it reverts, though. Can add that.

Reply:
Ah that is true - I miss that we add that modifier to `_cacheTokenRateIfNecessary`. Nice.